netdisco/xiaomi_gw.py

```diff
@@ -0,0 +1,68 @@
+"""Xiaomi Mi Home (aka Lumi) Gateway discovery."""
+import socket
+import json
+from datetime import timedelta
+
+
+DISCOVERY_PORT = 4321
+DISCOVERY_ADDRESS = "224.0.0.50"
+DISCOVERY_PAYLOAD = b'{"cmd":"whois"}'
+DISCOVERY_TIMEOUT = timedelta(seconds=2)
+
+
+class XiaomiGw(object):
+    """Base class to discover Xiaomi Gateway devices."""
+
+    def __init__(self):
+        """Initialize the discovery."""
+        self.entries = []
+
+    def scan(self):
+        """Scan the network."""
+        self.update()
+
+    def all(self):
+        """Scan and return all found entries."""
+        self.scan()
+        return self.entries
+
+    def update(self):
+        """Scan network for devices."""
+        entries = []
+
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        sock.settimeout(DISCOVERY_TIMEOUT.seconds)
+        sock.sendto(DISCOVERY_PAYLOAD, (DISCOVERY_ADDRESS, DISCOVERY_PORT))
+
+        while True:
+            try:
+                data, (address, _) = sock.recvfrom(1024)
+                #print(data, address)
+                try:
+                    entry = json.loads(data.decode("utf-8"))
+                except:  # pylint: disable=bare-except
+                    continue
+                if entry.get("model") == "gateway":
```

Comment:
Would it be possible to include an example json returned by the device here?

Btw, the gateway (and other Xiaomi devices / devices accessible over Mi Home app, like Yeelight bulbs) are also discoverable over mDNS with _miio._udp). Generally those devices will also respond on port 54321 for discovery (device type detection unknown to me).

Reply:
For example 

```
-> {"cmd":"whois"}
<- {"cmd":"iam","port":"9898","sid":"286fffffffff","model":"gateway","ip":"192.168.130.29"}
```

Reply:
> Btw, the gateway (and other Xiaomi devices / devices accessible with the Mi Home app, like Yeelight bulbs) are also discoverable over mDNS with _miio._udp).

D'oh! Fairly speaking, I wanted to play with mDNS regarding that, but that slipped, partly because I'm not much familiar with mDNS, and would need to dig into it. So instead I just used info from https://github.com/louisZL/lumi-gateway-local-api/blob/master/device_discover.md#1-网关设备发现设备发现不加密 .

@rytilahti , would you submit an alternative PR, using mDNS? If not, I can look into it, but per above, it would take me some time to dig into it. But we surely better use existing generic discovery protocols rather than add support for adhoc vendor-specific ones.

Reply:
I'm not currently able to do that, however, maybe it would be best to extend  https://github.com/home-assistant/netdisco/blob/master/netdisco/discoverables/yeelight.py to detect the gateway too, and rename it miio.py? Here's how I have started to do it in python-mirobo: https://github.com/rytilahti/python-mirobo/blob/master/mirobo/discovery.py#L48

Reply:
Thanks for all the hints! I prepared mDNS-based patch at https://github.com/home-assistant/netdisco/pull/144 . Let's move further discussion there.